### DISCUSSION TITLE
"invalid zip data" attempting to open downloaded zip file

### DISCUSSION THREAD

**Discussion by garretwilson** [101arrowz/fflate#149] [Q&A] [answered]
I have a large file `foo.txt` which I have zipped into `foo.txt.zip` using the latest stable 7z 22.01 on Windows 10. I have a web site that was already working downloading the raw `foo.txt` directly and processing the contents. I want to retrofit that existing page so that it instead downloads `foo.txt.zip` and extracts the original `foo.txt` before processing it as before.

Trying to follow the sample code in the [readme](https://github.com/101arrowz/fflate), I have this:

```javascript
<script type="module">

import { unzipSync } from './js/fflate.js';

fetch("foo.txt.zip").then(response => {
  if (!response.ok) {
    throw new Error(`HTTP error: ${response.status}`);
  }
  return unzipSync(response.arrayBuffer());
}).then(pathDataMap => {
  //TODO find `foo.txt` in the map
  //TODO process the `foo.txt` as before
}
</script>
```

The readme instructions are a bit terse (it's not even clear the type of value that I should pass to `unzipSync()`, but I think that the returned `pathDataMap` will have a map of paths within the zip file, and the values will each be `Uint8Array` of the uncompressed file data.

Nevertheless I don't even get that far: `unzipSync()` emits an error:

> Uncaught (in promise) Error: invalid zip data
>    at unzipSync (fflate.js:2512:13)

Where did I go astray in my code?

**Comment by 101arrowz** [comment, marked as answer]
The README is meant to show that you need to pass a `Uint8Array`, sorry if that wasn't clear. Right now you're passing it a promise containing an ArrayBuffer; instead you'd need:

```js
unzipSync(new Uint8Array(await response.arrayBuffer()));
```

Alternatively if you want to process the data without blocking the main thread:

```js
const src = new Uint8Array(await response.arrayBuffer());
return new Promise((resolve, reject) => {
  unzip(src, (err, data) => err ? reject(err) : resolve(data));
});
```

**Reply by garretwilson** [reply]
Note that as I'm inside the promise `.then()`, my browser doesn't let me use `await`. (I need to get caught up to speed on await.) But I can just return the `response.arrayBuffer()` and process it in the next chained `.then()` clause.

**Comment by garretwilson** [comment]
OK, thanks; I'll try that. I understand the synchronous methods might not be as fast, but my first step was just to get it functioning.

Regarding asynchronous process, I would eventually like to process the data asynchronously line by line in the decompressed data. I would want my method to be called asynchronously as data is decompressed Something like this (pseudocode):

```javascript
onMoreData(chunk) {
  //append chunk to buffer
  while(found '\n' in buffer) {
    //convert from UTF-8 to string
    //process line of data
  }
  //leave partial line in buffer for next time
}
```

Ideally as the Zip file is downloaded over the wire, new chunks are continually sent to fflate and fflate decompresses the chunks concurrently and sends them to `onMoreData()`. Is that even possible? (I see in the readme that I can "push" new chunks to the async streams, but if I've downloaded the entire Zip file there's no point in my pushing chunks. I would prefer to let the fetch API "push" data as they are available over the wire.)

This last part is probably best asked as a separate discussion question, but I couldn't resist asking when you mention the async stuff.

Thanks again for the response. I'll go try to get the basic sync version working first.

**Reply by 101arrowz** [reply]
What you're asking for is possible but quite pointless unless your file is dozens of megabytes. You need to adapt `fetch`'s ReadableStream interface to fflate's more imperative style, I can send some sample code in a few hours. For now, probably read up on Web Streams if you aren't familiar with them yet.

**Comment by garretwilson** [comment]
> What you're asking for is possible but quite pointless unless your file is dozens of megabytes.

My current compressed file is 40MB and the uncompressed is well over 200MB, and expected to grow. (I wish I could use LZMA2; 7z gets me down to 34MB. I'll bet Brotli would be better than Zip, so I'll look into that as well.)

Yes, I know I "shouldn't be doing it that way" if you think I'm writing an end-user application. (API, pull on demand, etc.) But this is a proof of concept being sent out to a few testers; waiting for a 3XMB download is feasible, but 200MB is not. And it would be poor decision-making to spend weeks writing a whole back-end, multiple client modules for pulling when needed, divide up the data, etc. before I even know for sure project is viable or useful.

So I just wanted to pull down an XXMB file and process the original XXXMB contents after decompressing it. I would have thought that would have been solved 10 years ago, but apparently we [have to wait for this to be built in](https://developer.mozilla.org/en-US/docs/Web/API/DecompressionStream). But I digress.

> I can send some sample code in a few hours.

Oh, I don't want to ask you to go write code. Writing code is something I can do. I just need a bit of direction and clarification sometimes.

If you have extra time you might document the API, though. 😉

**Reply by 101arrowz** [reply]
I would probably use GZIP over ZIP, it's the same compression algorithm but a lot less finicky to parse in a streaming way. If you want to stick with ZIP, you can read the README for an example of how to use `Unzip` (the streaming unzipper) and the [API documentation](https://github.com/101arrowz/fflate/blob/master/docs/classes/unzip.md).

**Reply by garretwilson** [reply]
Oh, there are API docs. ❤️ I had missed the link. Thank you thank you!

**Comment by garretwilson** [comment]
I wound up going with Brotli for this proof-of-concept.

Here are the data sizes:

* Original data: ~228MB
* Zip compressed (via 7z): ~41MB
* LZMA2 compressed (via 7z): 34MB
* Brotli compressed (via `brotli.exe`): ~33MB

Decompressing this using the ~150KB `decode.min.js` distributed with Google's Brotli repository takes 1472ms on my computer. I haven't tested, but I have no reason to doubt that fflate might be faster. Nevertheless, even if fflate took zero seconds to decompress, it would only save me less than1.5 seconds (already a small number), and I would imagine that downloading the extra 8MB in the Zip file would eat up much of those savings.
